### 소스코드 : https://wikidocs.net/81048

# 00. 개발환경 및 가상환경생성

## 1. 파이썬 설치하기

* www.python.org 

## 2. Flask 개발환경 준비하기

### 파이썬 가상환경

>* pip show jupyter notebook
>  - pip install jupyter notebook
>* python -m pip install --upgrade pip
>* pip -V
>* conda env list
>* conda create -n flask python=3.9.12
>* activate flask
>* python -m ipykernel install --user --name flask --display-name flask
>* conda install -n flask Flask
>* conda list -n flask Flask
>* conda remove -name flask --all
>* jupyter kernelspec list
>* jupyter kernelspec uninstall flask


# 01. 플라스크 개발 준비

## 01-1 플라스크 시작하기

* Flask는 2004년 오스트리아의 오픈소스 개발자 `아르민 로나허 Armin Rohacher가 만든 웹프레임워크`
* 플라스크는 아르민 로나허가 만우절에 장난삼아 던진 아이디어 였는데 관심이 높아져 서비스로 만들었다고 한다.
* `Flask는 Django와 더불어 파이썬 웹 프레임워크의 양대 산맥`으로 자리매김하고 있다.

><b>플라스크는 마이크로 웹프레임워크이다.</b>
>* Flask는 `마이크로 웹프레임워크`라고 부른다. 여기서 마이크로는 `한 개의 파이썬 파일로 작성`할 수 있다 
>* 또는 기능이 부족하다의미가 아니라 `프레임워크를 간결하게 유지하고 확장할 수 있도록 만들었다는 뜻`이다.

><b>확장성 있는 설계란?</b>
>* Flask에는 폼 Form, 데이터베이스 DataBase를 처리하는 기능이 없다.
>* Django는 폼과 데이터베이스를 처리하는 기능을 포함하고 있다. 쉽게 말해 덩치가 큰 프레임워크이다.
>* Flask는 확장 모듈이라는 것을 사용하여 이를 보완한다.
>* 이 의미는 처음부터 모든 기능을 포함하고 있지 않고 개발자가 필요에 따라 확장 모듈을 포함시켜가며 개발하면 된다.

><b>플라스크는 자유로운 프레임워크다.</b>
>* Flask는 한마디로 자유도가 높은 프레임워크이다.
>* 다른 프레이워크와는 달리 플라스크는 최소한의 규칙만 있기 때문에 개발의 자유도는 다른 프레임워크보다 높다.

In [1]:
!python -V
!conda env list
!conda list -n flask

Python 3.11.5
# conda environments:
#
base                  *  C:\Anaconda3




EnvironmentLocationNotFound: Not a conda environment: C:\Users\gilbaek\.conda\envs\flask



In [1]:
from flask import Flask, request
from markupsafe import escape

app = Flask(__name__)

@app.route('/')
def hello():
    return 'Hello, World!!'
    
if __name__ == "__main__":
    app.run()
    
# 이처럼 플라스크를 이용하면 파일 하나로 구성된 짧은 코드만으로 완벽하게 동작하는 웹프로그램을 만들 수 있다.

 * Serving Flask app '__main__'
 * Debug mode: off


 * Running on http://127.0.0.1:5000
Press CTRL+C to quit
127.0.0.1 - - [07/Nov/2023 17:48:13] "GET / HTTP/1.1" 200 -
127.0.0.1 - - [07/Nov/2023 17:48:13] "GET /favicon.ico HTTP/1.1" 404 -


## 01-2 플라스크 프로젝트 생성하기

### 1. 프로젝트 디렉토리 생성하기
- %cd d:\##.mydoc\03.flask\00.강의자료
- %ls
- %mkdir  d:\##.mydoc\03.flask\myproject
- %ls
- %pwd 

In [16]:
%cd d:\##.mydoc\03.flask\00.강의자료
%ls
%mkdir  d:\##.mydoc\03.flask\myproject
%ls
%pwd

d:\##.mydoc\03.flask\00.강의자료
 D 드라이브의 볼륨: 보조 디스크
 볼륨 일련 번호: 0CD1-E92B

 d:\##.mydoc\03.flask\00.강의자료 디렉터리

2022-08-27  오후 12:22    <DIR>          .
2022-08-27  오후 12:22    <DIR>          ..
2022-08-27  오전 09:19    <DIR>          .ipynb_checkpoints
2022-08-27  오후 12:22            14,899 01.플라스크 개발 준비.ipynb
2022-08-27  오후 12:18    <DIR>          img
               1개 파일              14,899 바이트
               4개 디렉터리  308,040,450,048 바이트 남음


하위 디렉터리 또는 파일 d:\##.mydoc\03.flask\myproject이(가) 이미 있습니다.


 D 드라이브의 볼륨: 보조 디스크
 볼륨 일련 번호: 0CD1-E92B

 d:\##.mydoc\03.flask\00.강의자료 디렉터리

2022-08-27  오후 12:22    <DIR>          .
2022-08-27  오후 12:22    <DIR>          ..
2022-08-27  오전 09:19    <DIR>          .ipynb_checkpoints
2022-08-27  오후 12:22            14,899 01.플라스크 개발 준비.ipynb
2022-08-27  오후 12:18    <DIR>          img
               1개 파일              14,899 바이트
               4개 디렉터리  308,040,450,048 바이트 남음


'd:\\##.mydoc\\03.flask\\00.강의자료'

In [17]:
%%writefile d:\##.mydoc\03.flask\myproject\pybo.py
from flask import Flask

#__name__이라는 변수에 'pybo' 문자열이 저장
app = Flask(__name__)

@app.route('/')
def hello():
    return 'Hello World!!!'cls


Overwriting d:\##.mydoc\03.flask\myproject\pybo.py


### 2. Appliation 실행

>`(flak) > flask run`

* Error
  - Could not locate a Flask application
  - You did not provide the "FLASK_APP" environment variable
  - 환경변수를 설정해야 한다.

<img src="./img/01.기본 애플리케이션 설정하기_01.png"/>

#### FLASK_APP 혼경 변수의 기본값

* 플라스크는 `FLASK_APP 환경변수가 지정되지 않은 경우 자동으로 app.py를 기본 애플리케이션으로 인식`한다.
* 따라서, pybo.py파일명을 app.py로 설정했다면 FLASK_APP환경변수를 별도로 지정하지 않아도 된다.
* 하지만 기본 애플리케이션을 pybo.py로 지정하려면 `set FLASK_APP=pybo`로 지정한 후 실행해야 한다.

>`(flaks) > set FLASK_APP=pybo`<br>
>`(flaks) > flask run`

<img src="./img/01.기본 애플리케이션 설정하기_02.png"/>

#### 웹브라우저 : http://127.0.0.1:5000

<img src="./img/01.기본 애플리케이션 설정하기_03.png"/>

#### 플라스크서버를 개발환경으로 실행

* 서버중단 `ctrl+c`후 개발환경으로 실행
* 플라스크가 `개발환경으로 설정하면 debug모드가 활성화`된다.
* 디버그모드는 오류가 발생하면 디버깅 결과 메시지를 웹브라우저에 출력해 준다.
* 또한, 프로그램을 변경하면 서버가 자동으로 다시 시작하여 변경된 내용을 적용해 준다.

>`(flaks) > set FLASK_ENV=develoment`<br>
>`(flaks) > flask run`

<img src="./img/01.기본 애플리케이션 설정하기_04_개발환경으로 실행.png"/>

#### 일괄명령 작성학

* pybo.cmd

In [19]:
%%writefile d:\##.mydoc\03.flask\myproject\pybo.cmd
@echo off
@cd d:\##.mydoc\03.flask\myproject  
@set FlASK_APP=pybo
@set FLASK_ENV=develoment
@conda activate flask

Overwriting d:\##.mydoc\03.flask\myproject\pybo.cmd


# 02. 플라스크 개발 기초 공사

## 02-1. 플라스크 기초 다지기

### 플라스크 프로젝트 구조

* `pybo.py파일은 pybo패키지로 변경`할 수 있다.
* `__init__.py`파일이 pybo.py를 대신할 것이다.

* `models.py` : `데이터베이스를 처리`하는 파일
  - ORM(Objeect Relation Mapping)을 지원하는 파이썬 데이터도구인 `SQLlchmey를 사용`한다.
  - SQLAlchemy는 모델기반 데이터베이스 처리
* `forms.py`  : 서버로 `전송되는 폼`을 처리하는 파일
  - 서버로 전송된 폼을 처리할 때 `WTForms 라이브러리를 사용`힌다.
* `views`     : `화면구성`하는 디렉토리
  - 여러가지 view파일을 저장
* `static`    : `css, js, image`저장 디렉토리
* `templates` : `html`파일 저정 디렉토리
* `config.py` : 프로젝트 `환경설정`파일
  - 프로젝트의 환경변수, 데이터베이스설정등의 환경설정을 이 파일에 저장

<img src="./img/02.플라스크 개발 기초_01.png"/>

## 02-2. 플라스크 애플리케이션 패토리

* `app = Flask(__name__)`에서 app는 Flask로 만든 객체를 의미한다.
* Flask는 app객체를 사용해 여러가지 설정을 진행한다. 
* 그런데 이와 같은 방식으로 app객체를 전역으로 사용하면 규모가 커질 수록 문제가 발생할 확률이 높아지다. 
* 순환참조 circular import 요류가 대표적이다.
* app객체를 전역으로 사용할 때 발생하는 문제를 예방하기 위해 `Application Factory를 사용`하는 것이다.

### 1단계 : pybo.py를 `__init.py__`로 변경

1. pybo.py -> __init__.py로 변경
1. myproject\pybo 폴더생성 후 __init__.py 이동

```dos
!mkdir ..\myproject\xxx
!move ..\myproject\__init__.py ..\myproject\xxx
!..\myproject\flask run
```
1. ../myproject>flask run
   - Flask기본앱을 `FLASK_APP=pybo`로 설정으로 인해 pybo모듈 즉, `pybo/__inin__.py`를 가리키게 된다.
   - 따라서, pybo.py파일을 __init__.p로 대체했지만 정상적으로 실행된다.
   
### 2단계 : Application Factory 사용하기

1. __init__.py수정

```python
from flask import Flask

def create_app():
    app = Flask(__name__)
    @app.route('/')
    def hello_pybo():
        return 'Application Factory - Hello World!!!'
    
    return app
```
1. create_app()함수가 app객체를 반환하도록 코드를 수정. 
1. 이때 app객체가 함수안에서 사용되므로 hello_pybo함수를 create_app 내부함수로 포함
1. 여기서 `create_app함수가 애플리케이션 팩토리다`
1. 서버를 재시작하면 정상적으로 실행된다.

In [ ]:
%mkdir ..\myproject\pybo\views

In [13]:
%%writefile  ..\myproject\pybo\views\main_views.py
from flask import Blueprint

bp = Blueprint('main', __name__, url_prefix="/")
@bp.route('/')
def hello_pybo():
    return 'Bluepring- Hello Wrold!!!'

Writing ..\myproject\pybo\views\main_views.py


## 02-3. 블루프린트로 라우트 함수 관리하기

* @app.route('/')와 같은 `애너테이션으로 매핑되는 함수를 라우트함수`라 한다.
* 새로운 URL이 생실 때 라우트함수를 create_app함수안에 계속 추가해야 하는 불편함이 있다.
* `이때 사용할 수 있는 클래스가 블루프린트 blue print`이다.

### 1단계 : 블루프린트 생성

1. 폴더 & 파일생성 pybo\views\main_views.py

>`%mkdir d:##.mydoc\03.flask\myproject\pybo\views`

```python
%%writefile  ..\myproject\pybo\views\main_views.py
from flask import Blueprint

bp = Blueprint('main', __name__, url_prefix="/")
@bp.route('/')
def hello_pybo():
    return 'Blueprint - Hello Wrold!!!'
```

1. 상기 코드는 __init__.py를 main_views.py로 변경
1. 단 애너테이션이 @app.route에서 `@bp.route로 변경, bp는 Blueprint객체이다.`

### 2단계 : 플라스크 앱 생성시 블루프린트 적용하기

1. __init__.py 변경

```python
from flask import Flask

def create_app():
    app = Flask(__name__)
    from .views import main_views
    app.register_blueprint(main_views.bp)    
    return app
```

2. main_views.py 변경

```python
from flask import Blueprint

bp = Blueprint('main', __name__, url_prefix="/")

@bp.route('/')
def index():
    return '<h1>Index</h1>'

@bp.route('/hello')
def hello():    
    return 'h3>Blueprint - Hello Wrold!!!</h3>'    
```

## 02-4. 모델로 데이터 처리하기

* `ORM Object Relation Mapping`을 이용하면 파이썬문법만으로도 DB를 처리할 수 있다.

### 1. 플라스크 ORM 라이브러리 설치

* 파이썬 ORM라이브러리중 가장 많이 `SQLAlchemy`를 사용한다.
* 파이썬 모델을 이용해 테이블을 생성하고 컬럼을 추가등의 작업을 할 수 있게하는 `Flask-Migrate`를 설치
* ORM라이브러리 설치하기
  - flask가상환경에서 실행
  - `pip install Flask-Migrate`
  - Flask-Migrate를 설치하면 SQLAlchemy도 함께 설치된다.
 
### 2. 설정파일 추가하기

* myproject\config.py 작성
  - SQLALCHEMY_DATABASE_URI : DB접속주소
  - SQLALCHEMY_TRACK_MODIFICATIONS : SQLAlchemy의 이벤트 처리 옵션

```python
# SQLite
import os

BASE_DIR = os.path.dirname(__file__)

SQLALCHEMY_DATABASE_URI = 'sqlite:///{}'.format(os.path.join(BASE_DIR, 'pybo.db'))
SQLALCHEMY_TRACK_MODIFICATIONS = False
```
```python
# mariadb
# create database flask
db = {
    'user': 'root',
    'password': '12345',
    'host': 'localhost',
    'port': 3306,
    'database': 'flask'
}

db_url = f"mariadb+mariadbconnector://{db['user']}:{db['password']}@{db['host']}:{db['port']}/{db['database']}?charset=utf8"
```
* mariadb connector 설치
```cmd
pip install mariadb
```
### 3. ORM적용

* init.py파일 수정해서 SQLAlchemy를 적용
  - app.config.from_object(config) : config.py정보를 호딩
  - 전역변수로 db, migrate객체를 생성
  - db, migrate객체를 지역변수로 사용하면 블루프린트 같은 다른 모듈에서 사용할 수 없다. 
  - init_app메서드를 이용해서 객체를 초기화

```python
from flask import Flask
from flask_migrate import Migrate
from flask_sqlalchemy import SQLAlchemy
import config
    
db = SQLAlchemy()
migrate = Migrate()

def create_app():
    app = Flask(__name__)
    app.config.from_object(config)
    
    # ORM
    db.init_app(app)
    migrate.init_app(app, db)
    
    # blueprint
    from .views import main_views
    app.register_blueprint(main_views.bp)    
    
    return app
```

### 4. 데이터베이스 초기화

* `flask db init`
  - 이 명령은 DB를 관리하는 `초기파일을 migrations라는 디렉토리에 자동을 생성`
  - `최초 한 번만 수행`하면 된다.
  
<img src="./img/02.플라스크 개발 기초_02_DB생성.png"/>

#### DB 관리 명령어

|명령어|설정|
|:----|:-----|
|flask db init|DB 초기화|
|flask db migrate|모델을 새로 생성하거나 변경|
|flask db upgrade|모델의 변경내용을 실제 DB에 적용|

### 5. 모델만들기

#### 1. 모델속성 구성하기

* question table
  - id      : 번호 PK
  - subject : 제목
  - content : 내용
  - create_date : 작성일자
  
* answertable
  - id
  - question_id
  - content
  - create_date
  
#### 2. 모델생성하기

* models.py
  - `db는 init.py에서 생성한 SQLAlchey객채`
  - db.relationship은 참조할 모델(Question)을 정의하고 backref는 역참조설정
    - 역참조설정이란 Question에서 Answer를 참조하는 것을 의미
    - 질문에 여러개의 답변이 달릴 수 있는데 역참조는 이 질문에 달린 답변을 참조할 수 있게 한다.
    - 질문에 해당하는 객체가 a_question이라면 a_question.answer_set와 같은 코드로 해당질문의 답변을 참조

```python
from pybo import db

class Question(db.Model):
    id = db.Column(db.Integer, primary_key=True)
    subject = db.Column(db.String(200), nullable=False)    
    content = db.Column(db.Text(), nullable=False)   
    create_date = db.Column(db.DateTime(), nullable=False)  
    
class Answer(db.Model):
    id = db.Column(db.Integer, primary_key=True)
    question_id = db.Column(db.Integer, db.ForeignKey('question.id', ondelete='CASCADE'))
    question = db.relationship('Question', backref=db.backref('answer_set'))
    content = db.Column(db.Text(), nullable=False)
    create_date = db.Column(db.DateTime(), nullable=False)     
```    

#### 3. 모델을 이용하여 Table생성하기

##### 3.1 모델가져오기

* init.py 수정

```python
    # 생략
    db.init_app(app)
    migrate.init_app(app, db)
    
    from . import models
    
    # 생략
```

##### 3.2 DB변경을 위한 리비전파일 생성하기

* DB변경사항 반영 : `flask db migrate` 수행

<img src="./img/02.플라스크 개발 기초_03_리비전파일.png"/>

##### 3.3 DB갱신하기

* DB갱신하기 : `flask db upgrade`

<img src="./img/02.플라스크 개발 기초_04_db upgrade.png"/>

* Table 생성확인하기

#### 4. 모델사용하기

##### 4.1 flask shell사용하기

###### 4.1.1 insert / select

* 플라스크 셀은 명령프롬프트에서 `flask shell`명령으로 실행한다.

```python
from pybo.models import Question, Answer
from datetime import datetime

q = Question(subject="제목", content="내용", create_date=datetime.now())

from pybo import db
db.session.add(q)
db.session.commit()
q.id

q = Question(subject='제목-두 번째 질문', content="두 번쨰 질문 내용....", create_date=datetime.now())
db.session.add(q)
db.session.commit()
q.id

Question.query.all()

Question.query.filter(Question.id==1).all()

Question.query.get(1)

Question.query.filter(Question.subject.like('%제목%').all()
```
<img src="./img/02.플라스크 개발 기초_06_insert.png"/>

###### 4.1.2 update

```python
q = Question.query.get(2)
q.subject = "[수정]제목"
db.session.commit()

q = Question.query.get(2)
db.session.delete(q)
db.session.commit()
```
###### 4.1.3 Answer
* Question과 Answer 모델이 연결된 상태
* Answer모델의 question속성에 역참조설정 backref=db.backref('answer_set')을 사용하면 관계를 쉽게 획득가능

```python
q = Question.query.get(1)
a = Answer(question=q, content="답변내용입니다.....", create_date=datetime.now())
db.session.add(a)
db.session.commit()

a.question
q.answer_set
```

## 02-5. 상세조회기능

###  1. 질문목록조회기능

#### 1.1 게시판 질문 목록 출력하기

1. main_views.py 수정
  - render_template함수는 템플릿파일을 화면에 출력한다.
  - 질문목록을 템플릿으로 전달하ㅕㄴ 전달 받은 데이터를 화면에 출력한다.

```pytnon
from flask import Blueprint, render_template
from pybo.models import Question

bp = Blueprint('main', __name__, url_prefix="/")

@bp.route('/')
def index():
    question_list = Question.query.order_by(Question.create_date.desc())
    return render_template('question/question_list.html', question_list=question_list)

@bp.route('/hello')
def hello():    
    return '<h3>Blueprint - Hello Wrold!!!</h3>'  
```

2. pybo/templates/question/questionquestion_list.html

```html
{% if question_list %}
    <ul>
        {% for question in question_list %}
            <li><a href="./detail/{{ question.id }}/">{{ question.subject }}</a></li>
        {% endfor%}
    </ul>
{% else %}
    <p>질문이 없습니다</p>
{% endif %}
```

#### 1.2 게시판 상세조회

1. 라우트함수 구현하기

* main_views.py 추가

```python
@bp.route('/detail/<int:question_id>/')
def detail(question_id):
    question = Question.query.get(question_id)
    return render_template('question/question_detail.html', question=question) 
```

2. question_detail.html

```html
<h1>{{ question.subject }}</h1>
<div>
    {{ question.content }}
</div>
```

3. 404 페이지

```python
@bp.route('/detail/<int:question_id>/')
def detail(question_id):
    # question = Question.query.get(question_id)
    question = Question.query.get_or_404(question_id)
    return render_template('question/question_detail.html', question=question) 
```

#### 1.3 blueprint로 기능 분리하기

1. 목록, 상세 조회기능 분리

* question_views.py

```python
from flask import Blueprint, render_template
from pybo.models import Question

bp = Blueprint('question', __name__, url_prefix="/question")

@bp.route('/list/')
def _list():
    question_list = Question.query.order_by(Question.create_date.desc())
    return render_template('question/question_list.html', question_list=question_list)

@bp.route('/detail/<int:question_id>/')
def detail(question_id):
    # question = Question.query.get(question_id)
    question = Question.query.get_or_404(question_id)
    return render_template('question/question_detail.html', question=question) 
```
* __init__.py

```python
from flask import Flask
from flask_migrate import Migrate
from flask_sqlalchemy import SQLAlchemy
import config
    
db = SQLAlchemy()
migrate = Migrate()

def create_app():
    app = Flask(__name__)
    app.config.from_object(config)
    
    # ORM
    db.init_app(app)
    migrate.init_app(app, db)
    
    from . import models
    
    # blueprint
    from .views import main_views, question_views
    app.register_blueprint(main_views.bp)    
    app.register_blueprint(question_views.bp) 
    
    return app
```
2. url_for로 리다이렉트 기능 추가

* main_views.py

```python
from flask import Blueprint, url_for
from werkzeug.utils import redirect

bp = Blueprint('main', __name__, url_prefix="/")

@bp.route('/')
def index():
    return redirect(url_for('question._list'))

@bp.route('/hello')
def hello():    
    return '<h3>Blueprint - Hello Wrold!!!</h3>'   
```

3. 하드코딩된 URL에 url_for함수 이용하기

* question_list.html

```html
{% if question_list %}
    <ul>
        {% for question in question_list %}
            <!-- <li><a href="./detail/{{ question.id }}/">{{ question.subject }}</a></li> -->
        <li><a href="{{ url_for('question.detail', question_id=question.id) }}">{{ question.subject }}</a></li>
        {% endfor%}
    </ul>
{% else %}
    <p>질문이 없습니다</p>
{% endif %}
```

## 02-6. 답변등록기능

### 1 답변데이터 저장하고 표시하기

#### 1.1 답변등록버튼만들기

* question_detail.html

```html
<h1>{{ question.subject }}</h1>
<div>
    {{ question.content }}
</div>
<h5>{{ question.answer_set|length }}개의 답변이 있습니다.</h5>
<div>
    <ul>
    {% for answer in question.answer_set %}
        <li>{{ answer.content }}</li>
    {% endfor %}
    </ul>
</div>
<form action="{{ url_for('answer.create', question_id=question.id) }}" method="post">
    <textarea name="content" id="content" rows="15"></textarea>
    <input type="submit" value="답변등록">
</form>
```

* 답변저장 URL은 form의 action속성에 지정된 `url_for('answer.create', question_id=question.id)`함수가 알려 준다.
* 답변등록버튼을 클릭하면 `POST방식으로 form의 action속성에 등록된 URL이 호출`된다.
* answer.create에 해당하는 URL을 찾을 수 없다는 메시지가 하면에 출력된다.
<img src="./img/02.플라스크 개발 기초_07_detail_error.png"/>
* 이 오류를 해결하려면 `Answer모델에 해당하는 블루프린트파일을 작성하고 create함수`를 만들어야 한다.

#### 1.2 답변 블루프린트 만들기

1. answer_view.py

```python
from datetime import datetime

from flask import Blueprint, url_for, request
from werkzeug.utils import redirect

from pybo import db
from pybo.models import Question, Answer

bp = Blueprint('answer', __name__, url_prefix='/answer') # answer_vewis.py파일이 answer이라는 이름의 BP파일임을 나타냄

@bp.route('/create/<int:question_id>', methods=('POST',)) # 답변저장템플릿의 form태그의 method
def create(question_id):
    question = Question.query.get_or_404(question_id)
    content = request.form['content']  # name속성이 content인 값
    answer = Answer(content=content, create_date=datetime.now())
    question.answer_set.append(answer) # 질문에 대한 답변들
    db.session.commit()
    return redirect(url_for('question.detail', question_id=question_id))
```

* create함수의 question_id는 URL에 전달된다.` http://localhost:5000/answer/create/1/를 요청 받으면 question_id에는 1이 전달`된다
* @bp.route('/create/<int:question_id>', methods=('POST',))에 post방식이 아니라면 아래와 같은 에러가 표시된다.

<img src="./img/02.플라스크 개발 기초_08_methodl_error.png"/>

* request.form['content']는 post방식으로 전송된 항목중 name속성이 content태그의 값을 리턴한다.
* question.answer_set는 질문에 달린 답변들을 의미한다.
  - Question과 Answer모델이 연결되어 있어 backref에 설정한 answer_set을 사용할 수 있다는 것을 기억하자
* `redirect의 question_id는 question_views.py파일에 있는 detail함수의 매개변수로 전달`된다.

#### 1.3 답변 블루프린트 적용하기

1. __init__.py

```python
from flask import Flask
from flask_migrate import Migrate
from flask_sqlalchemy import SQLAlchemy
import config
    
db = SQLAlchemy()
migrate = Migrate()

def create_app():
    app = Flask(__name__)
    app.config.from_object(config)
    
    # ORM
    db.init_app(app)
    migrate.init_app(app, db)
    
    from . import models
    
    # blueprint
    from .views import main_views, question_views, answer_views
    app.register_blueprint(main_views.bp)    
    app.register_blueprint(question_views.bp) 
    app.register_blueprint(answer_views.bp) 
    
    return app
```

* from .views import main_views, question_views, answer_views와 app.register_blueprint(answer_views.bp) 추가
* 실행이 성공적으로 완료가 되면 답변등록 화면이 표시가 된다.

<img src="./img/02.플라스크 개발 기초_09_답변등록화면.png"/>

#### 1.4 질문상셒이지에 답변표시하기

1. question_detail.html

```html
<h1>{{ question.subject }}</h1>

<div>
    {{ question.content }}
</div>

<h5>{{ question.answer_set|length }}개의 답변이 있습니다.</h5>
<div>
    <ul>
    {% for answer in question.answer_set %}
        <li>{{ answer.content }}</li>
    {% endfor %}
    </ul>
</div>

<form action="{{ url_for('answer.create', question_id=question.id) }}" method="post">
    <textarea name="content" id="content" rows="15"></textarea><br>
    <input type="submit" value="답변등록">
</form>
```

* {{ question.answer_set|length }}는 답변의 갯수를 의미
* `length는 템플릿 필터인데 객체의 길이를 리턴해 준다. |문자 뒤에 추가해서 사용`한다.
* 답변등록 성공화면이 표시

<img src="./img/02.플라스크 개발 기초_10_답변등록성공.png"/>

## 02-7. 화면 꾸미기

### 1. CSS적용하기

* css를 적용하려면 `css파일이 static디렉토리에 있어야 한다.` 이때 css파일은 플라스크에서 정적(static)파일로 분류한다.
* 정적파일은 주로 `이미지, 자바스크립, 스타일시트와 같은 파일을 의미`한다.

#### 1.1 static 디렉토리생성 및 css파일 작성

>`!mkdir ..\myproject\pybo\static`
```css
%%writefile ..\myproject\pybo\static\style.css
textarea {
    width:100%;
}
input[type=submit] {
    margin-top:10px;
}
```

#### 1.2 스타일시트 젹용하기

* question_detail.html

```html
<link rel="stylesheet" href="{{ url_for('static', filename='style.css') }}">
<h1>{{ question.subject }}</h1>

<div>
    {{ question.content }}
</div>

<h5>{{ question.answer_set|length }}개의 답변이 있습니다.</h5>
<div>
    <ul>
    {% for answer in question.answer_set %}
        <li>{{ answer.content }}</li>
    {% endfor %}
    </ul>
</div>

<form action="{{ url_for('answer.create', question_id=question.id) }}" method="post">
    <textarea name="content" id="content" rows="15"></textarea><br>
    <input type="submit" value="답변등록">
</form>
```

* `<link rel="stylesheet" href="{{ url_for('static', filename='style.css') }}">`
  - static디렉토리의 style.css파일을 연결
  - url_for()함수는 URL을 검색해주기 때문에 `정적파일의 URL도 찾아준다`

## 02-8. Bootstrap적용

### 1. Bootstrap적용하기

#### 1.1 Bootstrap 설치하기

1. https://getbootstrap.com/docs/5.2/getting-started/download/ 에서 다운로드
* 현재 `bootstrap-5.2.0-dist.zip`
* bootstrap.min.css파일을 static에 복사
<img src="./img/02.플라스크 개발 기초_11_bootstrap download.png"/>

#### 1.2 Bootstrap적용하기

1. question_list.html

```html
<link rel="stylesheet" href="{{ url_for('static', filename='bootstrap.min.css') }}">
<div class="container my-3">
    <table class="table">
        <thead>
        <tr class="table-dark">
            <th>번호</th>
            <th>제목</th>
            <th>작성일시</th>
        </tr>
        </thead>
        <tbody>
        {% if question_list %}
        {% for question in question_list %}
        <tr>
            <td>{{ loop.index }}</td>
            <td>
                <a href="{{ url_for('question.detail', question_id=question.id) }}">{{ question.subject }}</a>
            </td>
            <td>{{ question.create_date }}</td>
        </tr>
        {% endfor %}
        {% else %}
        <tr>
            <td colspan="3">질문이 없습니다.</td>
        </tr>
        {% endif %}
        </tbody>
    </table>
</div>
```

* `{{ loop.index }}`는 질문에 대한 번호(순서)를 의미

<h4 align="center">반복문의 loop속성</h4>

|loop속성|설명|
|:------|:------------------|
|loop.index|반복순서, 1부터 1씩증가|
|loop.index()|반복순서, 0부터 1씩증가|
|loop.first|반복순서가 첫 번째 순서이면 True 아니면 False|
|loop.last|반복순서가 마지막 순서이면 True 아니면 False|
    
#### 1.3 상세화면  Bootstrap적용하기

1. question_detail.html

```html
<link rel="stylesheet" href="{{ url_for('static', filename='bootstrap.min.css') }}">
<div class="container my-3">
    <!-- 질문 -->
    <h2 class="border-bottom py-2">{{ question.subject }}</h2>
    <div class="card my-3">
        <div class="card-body">
            <div class="card-text" style="white-space: pre-line;">{{ question.content }}</div>
            <div class="d-flex justify-content-end">
                <div class="badge bg-light text-dark p-2">
                    {{ question.create_date }}
                </div>
            </div>
        </div>
    </div>
    <!-- 답변 목록 -->
    <h5 class="border-bottom my-3 py-2">{{ question.answer_set|length }}개의 답변이 있습니다.</h5>
    {% for answer in question.answer_set %}
    <div class="card my-3">
        <div class="card-body">
            <div class="card-text" style="white-space: pre-line;">{{ answer.content }}</div>
            <div class="d-flex justify-content-end">
                <div class="badge bg-light text-dark p-2">
                    {{ answer.create_date }}
                </div>
            </div>
        </div>
    </div>
    {% endfor %}
    <!-- 답변 등록 -->
    <form action="{{ url_for('answer.create', question_id=question.id) }}" method="post" class="my-3">
        <div class="mb-3">
            <textarea name="content" id="content" class="form-control" rows="10"></textarea>
        </div>
        <input type="submit" value="답변등록" class="btn btn-primary">
    </form>
</div>
```

## 02-9. 표준HTML과 템플릿 상속

### 1. 템플릿을 표준 HTML구조로 변경

#### 1.1 템플릿 파일의 기본틀 작성

1. base.html
  - %%writefile  ..\myproject\pybo\templates\base.html
  - 탬플릿태그 : {% block content %} ... {% endblock %}

```html
<!doctype html>
<html lang="ko">
<head>
    <!-- Required meta tags -->
    <meta charset="utf-8">
    <meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
    <!-- Bootstrap CSS -->
    <link rel="stylesheet" href="{{ url_for('static', filename='bootstrap.min.css') }}">
    <!-- pybo CSS -->
    <link rel="stylesheet" href="{{ url_for('static', filename='style.css') }}">
    <title>Hello, pybo!</title>
</head>
<body>
<!-- 기본 템플릿 안에 삽입될 내용 Start -->
{% block content %}
{% endblock %}
<!-- 기본 템플릿 안에 삽입될 내용 End -->
</body>
</html>
```

#### 1.2 템플릿 파일 상속

* question_list.html
  - base.html 상속받기: {% extends 'base.html' %} 

```html
{% extends 'base.html' %}
{% block content %}
<div class="container my-3">
    <table class="table">
        <thead>
        <tr class="table-dark">
            <th>번호</th>
            <th>제목</th>
            <th>작성일시</th>
        </tr>
        </thead>
        <tbody>
        {% if question_list %}
        {% for question in question_list %}
        <tr>
            <td>{{ loop.index }}</td>
            <td>
                <a href="{{ url_for('question.detail', question_id=question.id) }}">{{ question.subject }}</a>
            </td>
            <td>{{ question.create_date }}</td>
        </tr>
        {% endfor %}
        {% else %}
        <tr>
            <td colspan="3">질문이 없습니다.</td>
        </tr>
        {% endif %}
        </tbody>
    </table>
</div>
{% endblock %}
```
#### 1.3 상세조회화면 수정

* question_detail.html

```html
{% extends 'base.html' %}
{% block content %}
<div class="container my-3">
    <!-- 질문 -->
    <h2 class="border-bottom py-2">{{ question.subject }}</h2>
    <div class="card my-3">
        <div class="card-body">
            <div class="card-text" style="white-space: pre-line;">{{ question.content }}</div>
            <div class="d-flex justify-content-end">
                <div class="badge bg-light text-dark p-2">
                    {{ question.create_date }}
                </div>
            </div>
        </div>
    </div>
    <!-- 답변 목록 -->
    <h5 class="border-bottom my-3 py-2">{{ question.answer_set|length }}개의 답변이 있습니다.</h5>
    {% for answer in question.answer_set %}
    <div class="card my-3">
        <div class="card-body">
            <div class="card-text" style="white-space: pre-line;">{{ answer.content }}</div>
            <div class="d-flex justify-content-end">
                <div class="badge bg-light text-dark p-2">
                    {{ answer.create_date }}
                </div>
            </div>
        </div>
    </div>
    {% endfor %}
    <!-- 답변 등록 -->
    <form action="{{ url_for('answer.create', question_id=question.id) }}" method="post" class="my-3">
        <div class="mb-3">
            <textarea name="content" id="content" class="form-control" rows="10"></textarea>
        </div>
        <input type="submit" value="답변등록" class="btn btn-primary">
    </form>
</div>
{% endblock %}
```

#### 1.4 스타일시트 비우기

* style.css

```css
/* 내용 전부 삭제 */
```

## 02-10. 폼모듈로 데이터검증하기

### 1. 플라스크폼 모듈 사용하기

#### 1.1 플라스크 폼 모듈 설치하기

* flask에서 폼을 사용하려면 `Flask-WTF를 설치`해야 한다
* 설치 : (flask) x:>`pip install Flask-WTF`
* Flask-WTK를 사용하려면 `플라스크 환경변수 SECRET_KEY가 필요`하다.
  - `SECRET_KEY는 CSRF(Cross-Site Request Forgery)라는 웹사이트 취약점 공격을 방지하는데 사용`한다.
* config.py 설정파일에 `SECRET_KEY="dev"`환경변수 추가
  - SECRET_KEY="dev"는 위험한 설정이다. 실제 서비스를 운영할 때 dev처럼 유추하기 쉬운 문자열을 사용하면 않된다.
* config.py

```python
# SQLite
# import os
# BASE_DIR = os.path.dirname(__file__)
# SQLALCHEMY_DATABASE_URI = 'sqlite:///{}'.format(os.path.join(BASE_DIR, 'pybo.db'))
# SQLALCHEMY_TRACK_MODIFICATIONS = False

# mariadb
db = {
    'user': 'root',
    'password': '12345',
    'host': 'localhost',
    'port': 3306,
    'database': 'flask'
}
# db_url = f"mariadb+mariadbconnector://{db['user']}:{db['password']}@{db['host']}:{db['port']}/{db['database']}?charset=utf8"
SQLALCHEMY_DATABASE_URI = f"mariadb+mariadbconnector://{db['user']}:{db['password']}@{db['host']}:{db['port']}/{db['database']}?charset=utf8"

# CSRF방지
SECRET_KEY="dev"
```
### 2. 질문등록 기능 만들기

#### 2.1 질문등록버튼

1. question_list.html

```html
(... 생략 ...)
    <table>
    </table>
    <a href="{{ url_for('question.create') }}" class="btn btn-primary">질문 등록하기</a>
```

2. question_view.py

```python
from flask import Blueprint, render_template
from pybo.models import Question
from ..forms import QuestionForm

bp = Blueprint('question', __name__, url_prefix="/question")

@bp.route('/list/')
def _list():
    question_list = Question.query.order_by(Question.create_date.desc())
    return render_template('question/question_list.html', question_list=question_list)

@bp.route('/detail/<int:question_id>/')
def detail(question_id):
    # question = Question.query.get(question_id)
    question = Question.query.get_or_404(question_id)
    return render_template('question/question_detail.html', question=question) 

@bp.route('/create/')
def create():
    form = QuestionForm()
    return render_template('question/question_form.html', form=form)
```

### 3. 질문등록 폼 클래스 만들기

* pybo\forms.py
  - QuestionForm과 같은 플라스크의 폼은 `FlaskForm 클래스를 상속`하여 만든다.
  - QuestionForm의 속성은 "제목"과 "내용"이다. 
  - 글자수의 제한이 있는 "제목"의 경우 StringField를 사용하고 글자수의 제한이 없는 "내용"은 `TextAreaField를 사용`한다. 
  - 참고 : https://wtforms.readthedocs.io/en/2.3.x/fields/#basic-fields
  - `StringField('제목', validators=[DataRequired()])` 에서 첫번째 입력인수인 "제목"은 폼 라벨(Label)이다
    - validators는 검증을 위해 사용되는 도구로 필수 항목인지를 체크하는 DataRequired, 
    - 이메일인지를 체크하는 Email, 길이를 체크하는 Length등이 있다. 
    - 참고 : https://wtforms.readthedocs.io/en/2.3.x/validators/#built-in-validators
    
```python
from flask_wtf import FlaskForm
from wtforms import StringField, TextAreaField
from wtforms.validators import DataRequired

class QuestionForm(FlaskForm):
    subject = StringField('제목', validators=[DataRequired()])
    content = TextAreaField('내용', validators=[DataRequired()])
```

### 4. 질문등록 템플릿 만들기

* question_form.html

```html
{% extends 'base.html' %}
{% block content %}
<div class="container">
    <h5 class="my-3 border-bottom pb-2">질문등록</h5>
    <form method="post" class="post-form my-3">
        
        {{ form.subject.label }}
        {{ form.subject()}}
        
        {{ form.content.label }}
        {{ form.content() }}
        
        <button type="submit" class="btn btn-primary">저장하기</button>
    </form>
</div>
{% endblock %}
```
<img src="./img/02.플라스크 개발 기초_12_질문등록하기.png"/>

* Method Error
  - 현재 폼이 POST 방식으로 데이터를 전송하기 때문
  - 현재 create 라우팅 함수에는 별도의 method 속성을 지정하지 않았으므로  GET 방식만 처리할 수 있다.
  
<img src="./img/02.플라스크 개발 기초_13_method_error.png"/>


### 5. 질문 전송방식 수정

* Method Error 해결 : question_vews.py 수정

```python
@bp.route('/create/', methods=('GET', 'POST'))
def create():
    form = QuestionForm()
    return render_template('question/question_form.html', form=form)
```

* 수정후 저장하기버튼을 클릭하면 에러화면은 보이지 않지만 별다른 반응이 없을 것이다.
* 왜냐하면 create함수에 데이터를 저장하는 코드를 작성하지 않았기 떄문이다.

### 6. 폼 데이터를 저장하는 코드 작성하기

* question_vews.py 수정

```python
from datetime import datetime

from flask import Blueprint, render_template, request, url_for
from werkzeug.utils import redirect
from .. import db

from pybo.models import Question
from ..forms import QuestionForm

bp = Blueprint('question', __name__, url_prefix="/question")

@bp.route('/list/')
def _list():
    question_list = Question.query.order_by(Question.create_date.desc())
    return render_template('question/question_list.html', question_list=question_list)

@bp.route('/detail/<int:question_id>/')
def detail(question_id):
    # question = Question.query.get(question_id)
    question = Question.query.get_or_404(question_id)
    return render_template('question/question_detail.html', question=question) 

@bp.route('/create/', methods=('GET', 'POST'))
def create():
    form = QuestionForm()
    if request.method == 'POST' and form.validate_on_submit():
        question = Question(subject=form.subject.data, content=form.content.data, 
            create_date=datetime.now())
        db.session.add(question)
        db.session.commit()
        return redirect(url_for('main.index'))    
    return render_template('question/question_form.html', form=form)
```
* request.method는 create 함수로 요청된 전송 방식을 의미
* form.validate_on_submit 함수는 전송된 폼 데이터의 정합성을 점검 
* 즉, DataRequired() 같은 점검 항목에 이상이 없는지 확인
* 즉, 위 코드에 추가한 내용은 POST 요청이고 폼 데이터에 이상이 없을 경우 
* 질문을 저장한 뒤 main.index 페이지로 이동하라는 내용
* 추가한 코드의 핵심은 데이터 전송 방식이 POST인지 GET인지에 따라서 달리 처리하는 부분이다. 
* 질문 목록에서 <질문 등록하기> 버튼을 누르거나 질문 등록 화면에서 <저장하기> 버튼을 누르면 
* 똑같이 localhost:5000/question/create/ 페이지를 요청하므로 create 함수가 이 요청을 받는다.
* 다만 create 함수에서 요청 방식을 구분해서 처리한다. 
* 즉, <질문 등록하기> 버튼을 누르는 것은 GET 방식 요청이므로 질문 등록 화면을 보여 주고, 
* <저장하기>버튼을 누르면 POST 방식 요청이므로 데이터베이스에 질문을 저장한다.
* 현재, 질문등록이 되지 않는다. 아래에서 해결예저

### 7. bootstrap 적용하기

1. question_form.html

```html
{% extends 'base.html' %}
{% block content %}
<div class="container">
    <h5 class="my-3 border-bottom pb-2">질문등록</h5>
    <form method="post" class="post-form my-3">
        
        {{ form.subject.label }}
        {{ form.subject(class="form-control") }}
        
        {{ form.content.label }}
        {{ form.content(class="form-control") }}
        
        <button type="submit" class="btn btn-primary">저장하기</button>
    </form>
</div>
{% endblock %}
```

### 8. 수작업으로 폼 작성하기

1. question_form.html

* {{ form.subject() }}처럼 폼을 생성하는 HTML 코드를 자동으로 생성하도록 했다. 
* 하지만 이 방식은 폼을 빠르게 만드는 데는 도움이 되지만 내가 원하는 디자인을 적용하기가 어렵다. 
* 이러한 단점을 보완하고자 이번에는 HTML을 직접 작성하는 방식으로 질문 등록 기능을 수정

```html
{% extends 'base.html' %}
{% block content %}
<div class="container">
    <h5 class="my-3 border-bottom pb-2">질문등록</h5>
    <form method="post" class="my-3">
        <div class="mb-3">
            <label for="subject">제목</label>
            <input type="text" class="form-control" name="subject" id="subject">
        </div>
        <div class="mb-3">
            <label for="content">내용</label>
            <textarea class="form-control" name="content" id="content" rows="10"></textarea>
        </div>
        <button type="submit" class="btn btn-primary">저장하기</button>
    </form>
</div>
{% endblock %}
```

### 3. 질문등록 기능 완성하기

* <저장하기>를 눌러도 화면에는 아무런 변화가 없었다. 
* 이번에는 그 이유를 알아보고 보완해서 질문 등록 기능을 완성해 보자.

#### 3.1 에러내용 표시하기

1. question_form.html

```html
{% extends 'base.html' %}
{% block content %}
<div class="container">
    <h5 class="my-3 border-bottom pb-2">질문등록</h5>
    <form method="post" class="post-form my-3">
        <!-- 오류표시 Start -->
        {% for field, errors in form.errors.items() %}
        <div class="alert alert-danger" role="alert">
            <strong>{{ form[field].label }}</strong>: {{ ', '.join(errors) }}
        </div>
        {% endfor %}
        <!-- 오류표시 End -->        
        <div class="form-group">
            <label for="subject">제목</label>
            <input type="text" class="form-control" name="subject" id="subject">
        </div>
        <div class="form-group">
            <label for="content">내용</label>
            <textarea class="form-control" name="content" id="content" rows="5"></textarea>
        </div>
        <button type="submit" class="btn btn-primary">저장하기</button>
    </form>
</div>
{% endblock %}
```
* 오류를 표시할 수 있는 영역을 위와 같이 추가했다. 
* create함수에서 form.validate_on_submit() 코드가 실패(false 반환)하면 폼에는 오류가 자동으로 등록된다. 
* 등록된 오류는 form.errors 속성을 사용하여 위와 같이 표시할 수 있다. 
* form.errors.items 의 field는 subject나 content와 같은 입력 폼의 필드를 의미한다.

<img src="./img/02.플라스크 개발 기초_14_오류표시.png"/>

#### 3.2 CSRF오류 처리

* form 태그 바로 밑에 `{{ form.csrf_token }} 코드를 삽입`하여 CSRF 토큰 오류 해결
* question_form.html 수정

<img src="./img/02.플라스크 개발 기초_15_질문등록성공.png"/>

#### 3.3 입력값유지및 오류메시지를 한글로 변경

1. 입력값 유지하기

   - <input ... name="subject" id="subject" value="{{ form.subject.data or '' }}">
   - <textarea ... name="content" id="content" rows="5" value="{{ form.subject.data or '' }}"></textarea>
   
```html
        <!-- (...생략...)       -->
        <div class="form-group">
            <label for="subject">제목</label>
            <input type="text" class="form-control" name="subject" id="subject" value="{{ form.subject.data or '' }}">
        </div>
        <div class="form-group">
            <label for="content">내용</label>
            <textarea class="form-control" name="content" id="content" rows="5" value="{{ form.subject.data or '' }}"></textarea>
        </div>
        <button type="submit" class="btn btn-primary">저장하기</button>
```
* `{{ form.subject.data or '' }}`를 입력하면 이미 전송한 데이터가 다시 설정된다. 
* {{ form.subject.data or '' }}에서 or ''은 "현재 템플릿이 GET 방식으로 요청되는 경우 
* 기존 입력값이 없으므로(None으로 출력) 이를 방지하기 위해서" 사용했다. 
* 즉, 이렇게 하면 form.subject.data에 값이 없을 때 None이 아니라 ''이 출력된다. 
* content필드에도 마찬가지 방법이 적용

2. 오류메시지 한글로 변경

* forms.py 수정

```python
from flask_wtf import FlaskForm
from wtforms import StringField, TextAreaField
from wtforms.validators import DataRequired

class QuestionForm(FlaskForm):
    # subject = StringField('제목', validators=[DataRequired()])
    # content = TextAreaField('내용', validators=[DataRequired()])
    subject = StringField('제목', validators=[DataRequired('제목은 필수입력 항목입니다.')])
    content = TextAreaField('내용', validators=[DataRequired('내용은 필수입력 항목입니다.')])
```

<img src="./img/02.플라스크 개발 기초_16_에러메시지한글표시.png"/>

### 4. 답변등록 기능 수정하기

1. 답변등록폼추가

* forms.py

```python
from flask_wtf import FlaskForm
from wtforms import StringField, TextAreaField
from wtforms.validators import DataRequired

class QuestionForm(FlaskForm):
    # subject = StringField('제목', validators=[DataRequired()])
    # content = TextAreaField('내용', validators=[DataRequired()])
    subject = StringField('제목', validators=[DataRequired('제목은 필수입력 항목입니다.')])
    content = TextAreaField('내용', validators=[DataRequired('내용은 필수입력 항목입니다.')])

class AnswerForm(FlaskForm):
    content = TextAreaField('내용', validators=[DataRequired('내용은 필수입력 항목입니다.')])
```

2. 답변등록 라우트 함수 수정

* answer_views.py

```python
from datetime import datetime
from flask import Blueprint, url_for, request, render_template
from werkzeug.utils import redirect

from .. import db
from ..forms import AnswerForm
from ..models import Question, Answer

bp = Blueprint('answer', __name__, url_prefix='/answer')


@bp.route('/create/<int:question_id>', methods=('POST',))
def create(question_id):
    form = AnswerForm()
    question = Question.query.get_or_404(question_id)
    if form.validate_on_submit():
        content = request.form['content']
        answer = Answer(content=content, create_date=datetime.now())
        question.answer_set.append(answer)
        db.session.commit()
        return redirect(url_for('question.detail', question_id=question_id))
    return render_template('question/question_detail.html', question=question, form=form)
```

3. CSRF코드와 오류표시 기능 추가

* question_detail.html

```html
{% extends 'base.html' %}
{% block content %}
<div class="container my-3">
    <!-- 질문 -->
    <h2 class="border-bottom py-2">{{ question.subject }}</h2>
    <div class="card my-3">
        <div class="card-body">
            <div class="card-text" style="white-space: pre-line;">{{ question.content }}</div>
            <div class="d-flex justify-content-end">
                <div class="badge bg-light text-dark p-2">
                    {{ question.create_date }}
                </div>
            </div>
        </div>
    </div>
    <!-- 답변 목록 -->
    <h5 class="border-bottom my-3 py-2">{{ question.answer_set|length }}개의 답변이 있습니다.</h5>
    {% for answer in question.answer_set %}
    <div class="card my-3">
        <div class="card-body">
            <div class="card-text" style="white-space: pre-line;">{{ answer.content }}</div>
            <div class="d-flex justify-content-end">
                <div class="badge bg-light text-dark p-2">
                    {{ answer.create_date }}
                </div>
            </div>
        </div>
    </div>
    {% endfor %}
    <!-- 답변 등록 -->
    <form action="{{ url_for('answer.create', question_id=question.id) }}" method="post" class="my-3">
        {{ form.csrf_token }}
        <!-- 오류표시 Start -->
        {% if form.errors %}
        <div class="alert alert-danger" role="alert">
            {% for field, errors in form.errors.items() %}
            <strong>{{ form[field].label }}</strong>
            <ul>
                {% for error in errors %}
                <li>{{ error }}</li>
                {% endfor %}
            </ul>
            {% endfor %}
        </div>
        {% endif %}
        <!-- 오류표시 End -->       
        <div class="mb-3">
            <textarea name="content" id="content" class="form-control" rows="10"></textarea>
        </div>
        <input type="submit" value="답변등록" class="btn btn-primary">
    </form>
</div>
{% endblock %}
```
4. 상세보기 라우트에 폼 적용하기

* question_view.py

```python
from datetime import datetime

from flask import Blueprint, render_template, request, url_for
from werkzeug.utils import redirect
from .. import db

from pybo.models import Question
from ..forms import QuestionForm, AnswerForm

bp = Blueprint('question', __name__, url_prefix="/question")

@bp.route('/list/')
def _list():
    question_list = Question.query.order_by(Question.create_date.desc())
    return render_template('question/question_list.html', question_list=question_list)

@bp.route('/detail/<int:question_id>/')
def detail(question_id):
    form = AnswerForm()
    question = Question.query.get_or_404(question_id)
    return render_template('question/question_detail.html', question=question,  form=form) 

@bp.route('/create/', methods=('GET', 'POST'))
def create():
    form = QuestionForm()
    if request.method == 'POST' and form.validate_on_submit():
        question = Question(subject=form.subject.data, content=form.content.data, 
            create_date=datetime.now())
        db.session.add(question)
        db.session.commit()
        return redirect(url_for('main.index'))    
    return render_template('question/question_form.html', form=form)
```